refactor: replace reflect.DeepEqual with assert.Equal

Surya-7890 opened the pull request:
This pr aims to replace reflect.DeepEqual with assert.Equal (`github.com/stretchr/testify/assert`) in testing code in the tree.

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Fixes: #40562

joestringer commented on operator/pkg/ciliumidentity/cache_test.go:
```diff
@@ -175,12 +174,12 @@ func TestCIDStateThreadSafety(t *testing.T) {
+func validateCIDState(t assert.TestingT, state, expectedState *CIDState) error {
+	if !assert.Equal(t, expectedState.idToLabels, state.idToLabels) {
 		return fmt.Errorf("failed to validate the state, expected idToLabels %v, got %v", expectedState.idToLabels, state.idToLabels)
 	}
```
For cases like this, there is no need to do the conditional check any more.

```suggestion
	assert.Equal(t, expectedState.idToLabels, state.idToLabels)
```

It's possible to add additional arguments for a message, but IMO in this case the testify library will already format all the same information without specifying a message here.

The same feedback applies to the other examples in this PR, please change those too.

joestringer replied:
Per my other comment on the PR, maybe it does make sense to take in the context message from the parent function and pass that as an extra parameter here. For extra points, you could add extra context to the message in this call. For instance, in this case it is checking the `idToLabels` parameter. Perhaps the failure message could look like "cid 1 added, idToLabels does not match expected result" (but note the first part would be coming as an argument to this function).

Surya-7890 replied:
I passed it as extra parameters in the assert.Equal function in my recent commit. So, it provides a detailed message on the assert failure, but also shows the fields that differ? Thanks for the suggestion!

joestringer commented on operator/pkg/ciliumidentity/cache_test.go:
```diff
@@ -55,7 +54,7 @@ func TestCIDState(t *testing.T) {
 		}
 
-		assert.NoError(t, validateCIDState(state, expectedState), "cid 1 added")
+		assert.NoError(t, validateCIDState(t, state, expectedState), "cid 1 added")
```
If the child function is going to do the assert, then it doesn't make sense to do an assert here. This could instead be:

```suggestion
		validateCIDState(t, state, expectedState, "cid 1 added")
```

In that case, it would be useful to add an extra parameter to that function to take this "cid 1 added" message through to all of the asserts in that function.

Surya-7890 replied:
Made the changes, as suggested. PTAL whenever possible!

Surya-7890 committed:
refactor: replace reflect.DeepEqual with assert.Equal

Fixes: #40562

This commit replaces some occurrences of reflect.DeepEqual with assert.Equal.
The changes are made particularly in the operator/pkg/ciliumidentity directory.

Signed-off-by: surya <suryamariappan9108@gmail.com>

Surya-7890 committed:
refactor: replace reflect.DeepEqual with assert.Equal

Fixes:#40562

This commit replaces instances of reflect.DeepEqual with assert.Equal in the test suite.
Using assert.Equal provides more informative failure messages.

Signed-off-by: surya <suryamariappan9108@gmail.com>

Surya-7890 committed:
refactor: replace reflect.DeepEqual with assert.Equal

Fixes: #40562

This commit fixes ci-lint errors, caused by previous commit

Signed-off-by: surya <suryamariappan9108@gmail.com>

jrajahalme commented on operator/pkg/ciliumidentity/cache_test.go:
```diff
@@ -175,14 +173,9 @@ func TestCIDStateThreadSafety(t *testing.T) {
-		return fmt.Errorf("failed to validate the state, expected labelsToID %v, got %v", expectedState.labelsToID, state.labelsToID)
-	}
+func validateCIDState(t assert.TestingT, state, expectedState *CIDState, cidStatus string) error {
+	assert.Equal(t, expectedState.idToLabels, state.idToLabels, "%s, idToLabels does not match expected result", cidStatus)
```
nit: `t.Helper()` here could make error messages more readable?

Surya-7890 replied:
Working on it currently, will add a commit, resolving this

jrajahalme commented on cilium-cli/clustermesh/clustermesh_test.go:
```diff
@@ -18,23 +18,23 @@ import (
 	copy(bCopy, b)
 
-	return slices.EqualFunc(a, bCopy, func(m1, m2 map[string]any) bool {
+	found := false
```
`found` will be set to `true` once any element in `a` is found in `bCopy`, and never set back to `false` even if all the rest of the elements are mismatches.

jrajahalme replied:
Maybe move this `found` into the first loop and add an outer `equal := true` that is set to `false` if any one element is not `found`?

Surya-7890 replied:
Yes, will work on it now. Thanks for the correction!

jrajahalme commented on operator/auth/spire/client_test.go:
```diff
@@ -476,13 +476,11 @@ func Test_resolvedK8sService(t *testing.T) {
 			got, err := resolvedK8sService(t.Context(), tt.args.client, tt.args.address)
-			if tt.wantedErr != nil && (err == nil || !reflect.DeepEqual(err.Error(), tt.wantedErr.Error())) {
-				t.Errorf("resolvedK8sService() error = %v, wantErr %v", err, tt.wantedErr)
+			if tt.wantedErr != nil && err != nil {
```
not sure, but maybe we should require `err == nil` when `tt.wantedErr == nil`?

Surya-7890 replied:
Yes, either of the following conditions should be true, if `tt.wantedErr` is nil

1. err == nil
2. assert.Equal(t, tt.wantedErr.Error(), err.Error())

Will make the changes, as soon as possible. Thanks for pointing out!

Surya-7890 replied:
Should i add another condition, asserting err is nil if tt.WantedErr is nil?

jrajahalme commented on pkg/datapath/iptables/reconciler_test.go:
```diff
@@ -426,10 +425,8 @@ func assertIptablesState(current, expected desiredState) error {
-		return fmt.Errorf("expected proxies info to be %v, found %v",
-			expected.proxies, current.proxies)
+	if len(current.proxies) != 0 && len(expected.proxies) != 0 {
+		assert.Equal(t, expected.proxies, current.proxies)
```
not introduced here, but this also fails to check for equal even if expected len is non-zero if current proxies happens to be of zero len.

Surya-7890 replied:
So, you are suggesting to run the assertion, if the expected-proxies len is non-zero, even if the current-proxies len is zero? Please correct me if misunderstood your suggestion.

jrajahalme commented on pkg/envoy/xds/server_e2e_test.go:
```diff
@@ -42,36 +41,30 @@ var resources = []*envoy_config_route.RouteConfiguration{
-				return resourcesAny[i].String() < resourcesAny[j].String()
-			})
-			result = reflect.DeepEqual(response.Resources, resourcesAny)
+func responseCheck(t assert.TestingT, response *envoy_service_discovery.DiscoveryResponse,
```
`t` is a new parameter here, could we add a `t.Helper()` call here ?

Surya-7890 replied:
Same as above. Will be adding a commit, changing this

jrajahalme reviewed (changes requested):
Loop comparing slices seems wrong.
